src/stream/src/executor/over_window/over_partition.rs

```diff
@@ -382,43 +424,32 @@ impl<'a, S: StateStore> OverPartition<'a, S> {
         }
 
         loop {
-            // Terminateability: `extend_cache_leftward_by_n` and `extend_cache_rightward_by_n` keep
+            // TERMINATEABILITY: `extend_cache_leftward_by_n` and `extend_cache_rightward_by_n` keep
             // pushing the cache to the boundary of current partition. In these two methods, when
             // any side of boundary is reached, the sentinel key will be removed, so finally
-            // `self::find_affected_ranges` will return ranges without any sentinels.
-
-            let (left_reached_sentinel, right_reached_sentinel) = {
-                // SAFETY: Here we shortly borrow the range cache and turn the reference into a
-                // `'cache` one to bypass the borrow checker. This is safe because we only return
-                // the reference once we don't need to do any further mutation.
-                let cache_inner = unsafe { &*(self.range_cache.inner() as *const _) };
-                let ranges =
-                    self::find_affected_ranges(self.calls, DeltaBTreeMap::new(cache_inner, delta));
-                self.stats.lookup_count += 1;
-
-                if ranges.is_empty() {
-                    // no ranges affected, we're done
-                    return Ok((DeltaBTreeMap::new(cache_inner, delta), ranges));
-                }
+            // `Self::find_affected_ranges_readonly` will return `Ok`.
 
-                let left_reached_sentinel = ranges.first().unwrap().0.is_sentinel();
-                let right_reached_sentinel = ranges.last().unwrap().3.is_sentinel();
+            // SAFETY: Here we shortly borrow the range cache and turn the reference into a
+            // `'cache` one to bypass the borrow checker. This is safe because we only return
+            // the reference once we don't need to do any further mutation.
+            let cache_inner = unsafe { &*(self.range_cache.inner() as *const _) };
```

Comment:
Out of scope for this PR, since it was already present previously.

I think if we create an `into_parts` method, we can retrieve the cache reference from `DeltaBTreeMap`, and can continue to repopulate the cache using this method.

Reply:
It can be hard I think. The point here is that, we want to return references to `CacheKey` which lives as long as `range_cache`, but when one round of `find_affected_ranges_readonly` didn't find desired results, we need to drop the intermediate reference and mutate the cache (this will invalidate the reference if we wrongly use `unsafe`) and try again, repeat until we find the desired result. But the compiler don't know we are not intended to return the intermediate, it think the intermediate refs are also possible to be returned, thus prohibit mutation on range cache.

Maybe we can try your method later.